internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ratelimit
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/yarpc/internal/clock"
+)
+
+// Note: This file is inspired by:
+// https://github.com/prashantv/go-bench/blob/master/ratelimit
+
+// Clock is the minimum necessary interface to instantiate a throttle with a
+// clock or fake clock, compatible with clocks created using
+// github.com/andres-erbsen/clock or go.uber.org/yarpc/internal/clock.
+type Clock interface {
+	Now() time.Time
+}
+
+// Throttle is a rate limiter.
+type Throttle struct {
+	sync.Mutex
+
+	last       time.Time
+	sleepFor   time.Duration
+	perRequest time.Duration
+	maxSlack   time.Duration
+	clock      Clock
+}
+
+// Option is an option for a rate limiter constructor.
+type Option func(l *Throttle)
+
+// NewThrottle returns a Throttle that will limit to the given RPS.
```

Comment:
if we're talking about "RPS" my assumption is that there's an RPS input for a throttle, but I'm not sure what it is.